modules/testrail.py

```diff
@@ -0,0 +1,289 @@
+
+    def create_new_plan_entry(
+        self,
+        plan_id,
+        suite_id,
+        name=None,
+        description=None,
+        case_ids=None,
+        config_ids=None,
+        runs=None,
+    ):
+        payload = {
+            "suite_id": suite_id,
+            "name": name,
+            "description": description,
+            "include_all": bool(case_ids),
+        }
+        if payload.get("include_all"):
+            payload["case_ids"] = case_ids
+        if payload.get("config_ids"):
+            payload["config_ids"] = config_ids
+        if runs:
+            payload["runs"] = runs
+        return self.client.send_post(f"/add_plan_entry/{plan_id}", payload)
+
+    def matching_configs(self, testrail_project_id, config_group_id, config_name):
+        configs = self.client.send_get(f"/get_configs/{testrail_project_id}")
+        matching_group = next(c for c in configs if c.get("id") == config_group_id)
+        logging.info(f"matching group|| {matching_group}")
+        cfgs = [
+            c for c in matching_group.get("configs") if c.get("name") == config_name
+        ]
+        logging.info(f"cfgs {cfgs}")
+        return cfgs
+
+    def add_config(self, config_group_id, name):
+        return self.client.send_post(f"/add_config/{config_group_id}", {"name": name})
+
+    def get_run(self, run_id):
+        return self.client.send_get(f"/get_run/{run_id}")
+
+    def get_test_results(self, run_id):
+        results_rs_json = self.client.send_get(f"/get_tests/{run_id}")
+        return results_rs_json.get("tests")
+
+    def update_test_cases(
+        self,
+        testrail_project_id,
+        testrail_run_id,
+        testrail_suite_id,
+        test_case_ids=[],
+        status="passed",
+    ):
+        status_key = {"passed": 1, "failed": 5, "skipped": 3}
+        if not test_case_ids:
+            test_case_ids = [
+                test_case.get("id")
+                for test_case in self._get_test_cases(
+                    testrail_project_id, testrail_suite_id
+                )
+            ]
+        data = {
+            "results": [
+                {"case_id": test_case_id, "status_id": status_key.get(status)}
+                for test_case_id in test_case_ids
+            ]
+        }
+        return self._update_test_run_results(testrail_run_id, data)
+
+    # Private Methods
+
+    def _get_test_cases(self, testrail_project_id, testrail_test_suite_id):
+        return self.client.send_get(
+            f"get_cases/{testrail_project_id}&suite_id={testrail_test_suite_id}"
+        )
+
+    def _update_test_run_results(self, testrail_run_id, data):
+        return self.client.send_post(f"add_results_for_cases/{testrail_run_id}", data)
+
+    def _get_milestones(self, testrail_project_id):
+        milestone_obj = self.client.send_get(f"get_milestones/{testrail_project_id}")
+        return milestone_obj.get("milestones")
+
```

Comment:
i regret my decision to return the `milestones` array in this return here given the new filters added with the testrail version bump, which is why i no longer do that in my implementation and instead opt to specify the `milestones` array as a composition of the response when i need it instead, but do whatever works best for your use case.

in order for my recommendation to work for checking if the milestone exists in the comment above, you'll have to update your method here to something like this:

```python
def _get_milestones(self, project_id, **filters):
        """
        Retrieves milestones for a given project with optional filters.

        Args:
            project_id (int): The ID of the project.
            **filters: Arbitrary keyword arguments representing API filters.

        Available Filters:
            is_completed (bool or int):
                - Set to True or 1 to return completed milestones only.
                - Set to False or 0 to return open (active/upcoming) milestones only.

            is_started (bool or int):
                - Set to True or 1 to return started milestones only.
                - Set to False or 0 to return upcoming milestones only.

            limit (int):
                - The number of milestones the response should return.
                - The default response size is 250 milestones.

            offset (int):
                - Where to start counting the milestones from (the offset).
                - Used for pagination.

        Returns:
            dict: The API response containing milestones.
        """

        if not project_id:
            raise ValueError("Project ID must be provided.")

        # Base endpoint
        endpoint = f"get_milestones/{project_id}"

        # Process filters
        if filters:
            # Convert boolean values to integers (API expects 1 or 0)
            for key in ["is_completed", "is_started"]:
                if key in filters and isinstance(filters[key], bool):
                    filters[key] = int(filters[key])

            # Build query parameters
            query_params = "&".join(f"{key}={value}" for key, value in filters.items())
            endpoint = f"{endpoint}&{query_params}"

        # Make API call
        return self.client.send_get(endpoint)
```

i know it's a lot more code, but it's also more flexible now. 

the only thing i'd change about it looking back on it, is to add type hints so those doc strings aren't needed any more.

Reply:
OK, I just went ahead and pasted this method in--it doesn't really change my workflow, but if it's more future-proof I'm for it.